### DISCUSSION TITLE
I think I encountered a very random and rare DateTime bug

### DISCUSSION THREAD

**Discussion by ronheywood** [AutoFixture/AutoFixture#1467] [Q&A]
A unit test failed unexpectedly with a stack trace:
```
System.ArgumentException : The supplied DateTime represents an invalid time. For example, when the clock is adjusted forward, any time in the period that is skipped is invalid. Parameter name: dateTime at System.TimeZoneInfo.ConvertTime(DateTime dateTime, TimeZoneInfo sourceTimeZone, TimeZoneInfo destinationTimeZone, TimeZoneInfoOptions flags, CachedData cachedData) 
```

The Autofixture generated a simple DTO, The property was a nullable datetime. 

The system expected the timezone info to be LOCAL and attempted to convert it to UTC
The stack trace emerged from this code:
```
TimeZoneInfo.ConvertTimeToUtc(d.Value, timezoneInfo);
```

(the Timezone info passed was "GMT Standard Time")

This test has been run many times per day for over a year, We've never seen it fail. 

I had a quick look in the code and saw that the datetime is created from a random long given as the ticks
Given the default range of ticks is 4 years, We believe that we must have been astonishingly unlucky to uncover this.

I'm not entirely sure how Ticks can ever be invalid, even when generated from a random long, 
although I note from the documentation that the first two bytes denote the timezone 
https://github.com/dotnet/runtime/blob/5535e31a712343a63f5d7d796cd874e563e5ac14/src/libraries/System.Private.CoreLib/src/System/DateTime.cs#L21

I have started trying to craft a test to replicate the issue, but I find it a bit weird to reason with.

I'm not sure what random long between a range of 2 valid ticks could be invalid 
- I figure it must be something to do with the first 2 bytes but I don't know how that is calculated

**Comment by mycroes** [comment]
I was intrigued by your issue, so I spent a little time investigating. AutoFixture by default limits the range to 2 years before today to 2 years after today. Those logically can be expressed as ticks just fine, so there shouldn't be any issue there.

Now you're also talking about the internal time representation of DateTime. I looked at the linked page, and while they are using the first two *bits* (not bytes) for representing the `DateTimeKind`, that logically doesn't affect any DateTime in the allowed range of `DateTime.MinValue` (00:00:00.0000000 UTC, January 1, 0001) to `DateTime.MaxValue`  (23:59:59.9999999 UTC, December 31, 9999) and especially not any DateTime from 2 years ago to 2 years in the future. The range check is actually implemented by casting the `long ticks` to `ulong` and checking if it's over the number of ticks for 10.000 years. That means if either the first or second bit is set, it will definitely throw on this check, but that's an `ArgumentOutOfRangeException` instead of `ArgumentException`.

Unfortunately your 'stack trace' is only a single exception, the actual stack trace might point where this exception was coming from. Since you only had this issue once and noone else is reporting similar issues, I'd rather expect this to be a system issue than a software issue, or a software issue that's unrelated to AutoFixture (because mathematically, this all seems to be just fine). I wouldn't worry about this specific issue, I guess you'll notice far more random errors if there really is something wrong with your system and on the AutoFixture side this seems to be fine as well (and if it wasn't, it would give you a different exception).

**Comment by ronheywood** [comment]
I'd forgotten about this, thanks for your analysis
I vaguely remember it - it occurred in a build pipeline, and it's run many hundreds of times per day.

Haven't seen it since
Should probably just close https://github.com/AutoFixture/AutoFixture/discussions/1467 as "Unable to reproduce"
